src/basicfuns.jl

```diff
@@ -233,17 +233,12 @@ That is, `r` is overwritten with `exp.(x)`, normalized to sum to 1.
 See the [Wikipedia entry](https://en.wikipedia.org/wiki/Softmax_function)
 """
 function softmax!(r::AbstractArray{<:Real}, x::AbstractArray{<:Real})
-    n = length(x)
-    length(r) == n || throw(DimensionMismatch("Inconsistent array lengths."))
+    length(r) == length(x) || throw(DimensionMismatch("inconsistent array lengths"))
     u = maximum(x)
-    s = zero(eltype(r))
-    @inbounds for i = 1:n
-        s += (r[i] = exp(x[i] - u))
-    end
-    invs = inv(s)
-    @inbounds for i = 1:n
-        r[i] *= invs
+    map!(r, x) do xi
+        return exp(xi - u)
     end
+    LinearAlgebra.lmul!(inv(sum(r)), r)
```

Comment:
Isn't it slower to compute the sum again at the end? We have again the same issue that `CuArray`'s requirements are not necessarily the best for `Array`.

Reply:
I benchmarked many different implementations such as the current one, iteration with `eachindex`, `map!`, broadcasting (with `vec` to avoid dimension mismatches), with additional `sum` and without, and also more obscure ones such as `map!` closed over a summation variable. Both for arrays with 10 and 1000 elements the timings were quite similar and in most cases I didn't notice any clear performance regressions. Surprisingly, in some cases actually the explicit `sum` was faster - maybe it is easier e.g. for the compiler to optimize the for loop without manual summation?

So my impression was that the `sum` does not cause any clear performance regressions and therefore I went for this more general approach. The finding also didn't seem very surprising since intuitively exponentiation is the most expensive part of the softmax function.

Reply:
Cool, I guess the simple version is OK then.